src/mpi/init/mpir_init.c

```diff
@@ -131,8 +133,8 @@ int MPII_Init_thread(int *argc, char ***argv, int user_required, int *provided,
     MPID_Thread_init(&err);
     MPIR_Assert(err == 0);
 
-    mpi_errno = MPIR_T_env_init();
-    MPIR_ERR_CHECK(mpi_errno);
+    /* MPI_T was previously set up here, but that moved to MPII_qmpi_setup to support using
+     * environment variables before MPI_Init. */
```

Comment:
Could you explain why `MPIR_T_env_init()` need be moved away? Is it because now `MPII_qmpi_setup` has to be run before `MPI_Init`? Anyway, I think you can generally explain the flow a bit.

Reply:
We need the CVAR system to be set up before we even get into `MPII_qmpi_setup`. Let's look at this code from `MPI_Info_get` as an example:


```c
int MPI_Info_get(MPI_Info info, const char *key, int valuelen, char *value, int *flag)
{
    // ... snip ...

    /* Initialize the tools that have been registered */
    for (int i = 1; i < MPIR_QMPI_num_tools + 1; i++) { // CVARS must be available here
        MPIR_QMPI_tool_init_callbacks[i](i);
    }
    if (!MPIR_QMPI_is_initialized) {
       int mpi_errno = MPI_SUCCESS;
       mpi_errno = MPII_qmpi_setup();
       if (mpi_errno != MPI_SUCCESS) {
           return mpi_errno;
       }
    }

    MPII_qmpi_stash_first_tools();

    if (MPIR_QMPI_num_tools == 0)
        return QMPI_Info_get(context, 0, info, key, valuelen, value, flag);

    fn_ptr = (QMPI_Info_get_t *) MPIR_QMPI_first_fn_ptrs[MPI_INFO_GET_T];

    // The first chance MPICH has to call its own setup is here
    return (*fn_ptr) (context, MPIR_QMPI_first_tool_ids[MPI_INFO_GET_T], info, key, valuelen, value,
            flag);
}

int QMPI_Info_get(QMPI_Context context, int tool_id, MPI_Info info, const char *key, int valuelen,
                  char *value, int *flag)
{
    int mpi_errno = MPI_SUCCESS;
    MPIR_Info *info_ptr = NULL;

    // ... snip ...

    MPIR_Info_get_ptr(info, info_ptr);

    mpi_errno = MPIR_Info_get_impl(info_ptr, key, valuelen, value, flag);

    // ... snip ...
}
```

The workflow for a QMPI tool is:

1. The tool creates a callback function to be triggered *before* MPI is initialized.
2. The tool creates a constructor function (e.g., `__attribute__ ((constructor))`) and registers it with MPI via `QMPI_Register_tool_name`)
3. The user sets an environment variable to pick the tool(s) that it wants. 
4. The first time an MPI function is called, MPI calls all of the registered callback functions that match the name of the tool with the names in the environment variable

All of this has to happen *before* any other MPI function, because the tool needs to be able to intercept that function (in the same way they already can with PMPI).

Reply:
`MPIR_T_env_init` internally checks static state `initialized`, so it can be called multiple times. I think we should retain the lines in `MPII_Init_thread` rather than replace it with a comment.

Actually, I think `MPII_qmpi_setup` should do the same, i.e. move `MPIR_QMPI_is_initialized` into internal static variable and internally checks it rather than let caller call it. And it be better to have it called explicitly inside `MPII_Init_thread` as well. That way, people can see and understand the code rather than making wild guess what is happening.

Now, generally, the MPI functions QMPI need access are always inside `MPI_Init/Finalize`, so it is guaranteed qmpi is initialized, right?

And for functions that can be called before `MPI_Init`, I don't think it should be accessed by QMPI, since they can be called after `MPI_Finalize` as well, at which point QMPI can't be shutdown properly.

Reply:
Oh btw, `QMPI_Info_get` need be moved into `#ifndef MPI_FROM_PMPI` block to avoid double definition, right?

Reply:
> `MPIR_T_env_init` internally checks static state `initialized`, so it can be called multiple times. I think we should retain the lines in `MPII_Init_thread` rather than replace it with a comment.

I replaced these to make sure that things work with `--enable-qmpi` is not set. The difference was that I had this macro guarded with `#ifndef ENABLE_QMPI`.

> Actually, I think `MPII_qmpi_setup` should do the same, i.e. move `MPIR_QMPI_is_initialized` into internal static variable and internally checks it rather than let caller call it. And it be better to have it called explicitly inside `MPII_Init_thread` as well. That way, people can see and understand the code rather than making wild guess what is happening.

It doesn't seem that confusing to me as it is, but if you prefer it the other way, that's not a difficult change.

> Now, generally, the MPI functions QMPI need access are always inside `MPI_Init/Finalize`, so it is guaranteed qmpi is initialized, right?

No, there are some QMPI functions that need to be used outside of when MPI would be initialized (explicitly as in the world model or implicitly as in the sessions model). For instance, `QMPI_Register_tool_name`.

> And for functions that can be called before `MPI_Init`, I don't think it should be accessed by QMPI, since they can be called after `MPI_Finalize` as well, at which point QMPI can't be shutdown properly.

I'm not sure what a good solution is here. You're right that it's hard to tell when it's ok to safely clean up the internal QMPI information, particularly in light of sessions. Just because all of the sessions are finalized doesn't mean that MPICH can clean up all internal state. I think it's the same problem for QMPI as it is for everything else...

Reply:
> No, there are some QMPI functions that need to be used outside of when MPI would be initialized (explicitly as in the world model or implicitly as in the sessions model). For instance, QMPI_Register_tool_name.

Is there anything different in this regard compared with `MPI_T`?

Anyway, before further discussions, maybe you can clarify the items you agreed to work on and the issues we are still debating?

Reply:
> Oh btw, `QMPI_Info_get` need be moved into `#ifndef MPI_FROM_PMPI` block to avoid double definition, right?

I'm not sure about this one. Is there a way to force MPICH to not use weak symbols? If we just stick it inside that `#ifndef`, it wouldn't be defined for the weak symbols versions.

Reply:
> > Oh btw, `QMPI_Info_get` need be moved into `#ifndef MPI_FROM_PMPI` block to avoid double definition, right?
> 
> I'm not sure about this one. Is there a way to force MPICH to not use weak symbols? If we just stick it inside that `#ifndef`, it wouldn't be defined for the weak symbols versions.

On mac we always build `libmpi` and `libpmpi` separately. In `libmpi` it doesn't need see internal functions since it only calls `PMPI_xxx`. I guess currently even if you stick the internal function in both `libmpi` and `libpmpi`, the dynamic linker won't complain since it simply discards unused symbols, but it is not ideal.

You can reference the `polymorph` in `binding_c.py`, where we generate `MPII_Comm_get_attr` inside the `#ifndef MPI_FROM_PMPI` block. We can do something similar here.

You can just leave it if you are not sure, I can add a commit later to do just that.

Reply:
> Anyway, before further discussions, maybe you can clarify the items you agreed to work on and the issues we are still debating?

I'm, not entirely sure what you were asking. It was something to do with not allowing QMPI functions before init?

Everything I agree is good.

Reply:
> I'm, not entirely sure what you were asking. It was something to do with not allowing QMPI functions before init?
> 
> Everything I agree is good.

I am asking *if* there are things that you haven't agreed or had reservations about, re-iterate them. Sometimes it is difficult to tell which or whether you agreed or not 😉